src/colmap/scene/frame.h

```diff
@@ -0,0 +1,264 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "colmap/geometry/rigid3.h"
+#include "colmap/util/types.h"
+
+#include <map>
+#include <set>
+#include <vector>
+
+namespace colmap {
+
+// Sensor type
+enum class SensorType {
+  Camera = 0,
+  IMU = 1,
+  Location = 2,  // include GNSS, radios, compass, etc.
+};
+
+// Unique identifier of the sensor
+typedef std::pair<SensorType, uint32_t> sensor_t;
+
+// Unique identifier of the data point from a sensor
+typedef std::pair<sensor_t, uint64_t> data_t;
+
+// Rig calibration storing the sensor from rig transformation
+class RigCalibration {
+ public:
+  // Access the unique identifier of the rig
+  inline rig_t RigId() const;
+  inline void SetRigId(rig_t rig_id);
+
+  // Add sensor into the rig
+  // ``AddReferenceSensor`` needs to called first before all the ``AddSensor``
+  // operation
+  inline void AddReferenceSensor(sensor_t ref_sensor_id);
+  inline void AddSensor(sensor_t sensor_id,
+                        const Rigid3d& sensor_from_rig = Rigid3d(),
+                        bool is_fixed = false);
+
+  // Check whether the sensor exists in the rig
+  inline bool HasSensor(sensor_t sensor_id) const;
+
+  // Count the number of sensors available in the rig
+  inline size_t NumSensors() const;
+
+  // Access the reference sensor id (default to be the first added sensor)
+  inline sensor_t RefSensorId() const;
+
+  // Check if the sensor is the reference sensor of the rig
+  inline bool IsReference(sensor_t sensor_id) const;
+
+  // Access sensor from rig transformations
+  inline Rigid3d& SensorFromRig(sensor_t sensor_id);
+  inline const Rigid3d& SensorFromRig(sensor_t sensor_id) const;
+
+ private:
+  // Unique identifier of the device.
+  rig_t rig_id_;
+
+  // Reference sensor id which has the identity transformation to the rig.
+  sensor_t ref_sensor_id_;
+
+  // list of sensors
+  std::set<sensor_t> sensor_ids_;
+
+  // sensor_from_rig transformation.
+  std::map<sensor_t, Rigid3d> map_sensor_from_rig_;
+  std::map<sensor_t, bool> is_fixed_sensor_from_rig_;  // for optimization
+};
+
+class Frame {
+ public:
+  // Access the unique identifier of the frame
+  inline frame_t FrameId() const;
+  inline void SetFrameId(frame_t frame_id);
+
+  // Access data ids
```

Comment:
Sorry, I probably forgot and we discussed this some months ago. Not sure I fully understand what the data_t is, in particular the uint64_t refers to what?

Reply:
Sorry, still don't fully understand what this translates to in practice. So, in case of an image, what exactly would be the uint64_t?

Reply:
The uint64_t is ``image_id``. ``data_t`` is ``std::pair<std::pairSensorType, sensor_t>, uint64_t>``, where the ``sensor_t`` field corresponds to ``camera_id``, and ``uint64_t`` corresponds to ``image_id`` before.

Reply:
Got it, thanks.